Add String Tutorial

iUnknwn opened the pull request:
This adds an introduction to using strings in the Windows API, and within the windows crate.

iUnknwn committed:
Add String Tutorial

This adds an introduction to using strings in the Windows API, and
within the windows crate.

iUnknwn commented:
There is one `todo` left in the documentation - I'm not sure if the `s!` macro does any sort of conversion, so I marked that as a "todo" item.

iUnknwn committed:
Fix minor bug and clarify string length behavior.

Fixes a bug where the two-step string example included a null
terminator.

I'm a bit unsure of line 127, and if it's valid to set length
to the exact end of the string, or if that's invalid, because
that memory does contain a null terminator.

However, this should do a better job of expalining how the two-step
string allocation process works.

iUnknwn committed:
Adjust Vec Length + Add Notes on widestring

Adjust an earlier example so that the vec length is set to before
the null terminator. This doesn't have any cost, and is slightly
cleaner.

Additionally, add notes about using the widestring crate, which
can work with utf16 strings directly.

kennykerr commented:
Thanks! Will review in the morning.

kennykerr commented on src/summary.md:
```diff
@@ -20,3 +20,4 @@
 - [Standalone code generation](rust-getting-started/standalone-code-generation.md)
 - [Creating your first DLL in Rust](rust-getting-started/creating-your-first-dll.md)
 - [Implement a traditional Win32-style API](rust-getting-started/implement-win32-api.md)
+- [Working with Strings in the Windows Crate](rust-getting-started/string-tutorial.md)
```
```suggestion
- [Working with Strings in the Windows Crate](rust-getting-started/string-tutorial.md)
```

Just some capitalization inconsistencies - also in the actual topic. 
```suggestion
- [Working with strings in the windows crate](rust-getting-started/string-tutorial.md)
```

kennykerr commented on src/rust-getting-started/string-tutorial.md:
```diff
@@ -0,0 +1,238 @@
+# Working with Strings in the Windows Crate
```
```suggestion
# Working with strings in the windows crate
```

kennykerr commented on src/rust-getting-started/string-tutorial.md:
```diff
@@ -0,0 +1,238 @@
+
+Let's look at an example using the simple [`MessageBox` function](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/WindowsAndMessaging/fn.MessageBoxW.html), which displays a pop up dialog box. We'll use the wide version (`MessageBoxW`) for this example. 
+
+If you want to call a Windows API using a string literal, the `windows-strings` crate has macros to generate windows strings from string literals:
```
```suggestion
If you want to call a Windows API using a string literal, the `windows-strings` crate has macros to generate Windows strings from string literals:
```

kennykerr commented on src/rust-getting-started/string-tutorial.md:
```diff
@@ -0,0 +1,238 @@
+If you want to call a Windows API using a string literal, the `windows-strings` crate has macros to generate windows strings from string literals:
+  - `h!` generates an HSTRING from a string literal, adding a null terminator and converting to UTF-16.
+  - `w!` does the same, but generates a PCWSTR instead of an HSTRING. 
+  - `s!` generates a PCSTR with a null terminator. **TODO: does this do any conversion**
```
It just appends a null terminator.

kennykerr commented on src/rust-getting-started/string-tutorial.md:
````diff
@@ -0,0 +1,238 @@
+```rust
+fn show_message_box_u16_conversion(text: &str, caption: &str) -> MESSAGEBOX_RESULT {
+    // convert our text and caption to UTF-16 bytes
+    let mut text = text.encode_utf16().collect::<Vec<u16>>();
+    let mut caption = caption.encode_utf16().collect::<Vec<u16>>();
+
+    // add null terminators
+    text.push(0u16);
+    caption.push(0u16);
````
In defense of the Standard Library, this is typically simpler using encode/chain/collect.

kennykerr commented on src/rust-getting-started/string-tutorial.md:
````diff
@@ -0,0 +1,238 @@
+If we wanted to call the message box we could use the windows crate with the `Win32_UI_WindowsAndMessaging` feature, and call:
+
+```rust
+fn show_message_box_literals() -> MESSAGEBOX_RESULT {
````
Brevity can be helpful - none of these functions add any additional insights - its simpler and easier to read just to include the code snippets without the surrounding functions. The rustdoc compiler also suggests avoiding wrapper functions.

iUnknwn replied:
My thought was that the function signature gave a bit more details on the type of string being returned for the later functions, but I can definitely revise and add comments to provide the same information.

kennykerr commented on src/rust-getting-started/string-tutorial.md:
```diff
@@ -0,0 +1,238 @@
+# Working with Strings in the Windows Crate
+
+Calls involving strings can be tricky; there are several different types of strings in the Windows APIs:
```
Try to be as concise as possible. The fact that this topic exists is sufficient to indicate that strings can be tricky. 😊
```suggestion
There are several string types in the Windows APIs including:
```

iUnknwn committed:
Make edits requested in PR Review.

Includes a few capitalization and style changes as requested. Also:
  - Removes wrapping function from each code example
  - Removes todo in s! macro description

iUnknwn reviewed (commented):
I think all comments should be resolved. Please let me know if you need me to make any other changes.

kennykerr commented:
Thanks!
